torch.fft helper functions

peterbell10 opened the pull request:
Part of gh-42175. This implements the `torch.fft` helper functions: `fftfreq`, `rfftfreq`, `fftshift` and `ifftshift`.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45164 torch.fft: Two dimensional FFT functions
* #45163 codegen: Improve array default handing
* **#44877 torch.fft helper functions**

Differential Revision: [D24043473](https://our.internmc.facebook.com/intern/diff/D24043473)

peterbell10 committed:
WIP: torch.fft helper functions

fftfreq, rfftfreq, fftshift and ifftshift

[ghstack-poisoned]

peterbell10 committed:
Update on "WIP: torch.fft helper functions"


fftfreq, rfftfreq, fftshift and ifftshift

[ghstack-poisoned]

peterbell10 committed:
Update on "WIP: torch.fft helper functions"


fftfreq, rfftfreq, fftshift and ifftshift

[ghstack-poisoned]

peterbell10 committed:
Update on "WIP: torch.fft helper functions"


fftfreq, rfftfreq, fftshift and ifftshift

[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft helper functions"

Part of gh-42175. This implements the `torch.fft` helper functions: `fftfreq`, `rfftfreq`, `fftshift` and `ifftshift`.

* #43009 Cleanup tracer handling of optional arguments

[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft helper functions"

Part of gh-42175. This implements the `torch.fft` helper functions: `fftfreq`, `rfftfreq`, `fftshift` and `ifftshift`.

* #43009 Cleanup tracer handling of optional arguments

[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft helper functions"

Part of gh-42175. This implements the `torch.fft` helper functions: `fftfreq`, `rfftfreq`, `fftshift` and `ifftshift`.

* #43009 Cleanup tracer handling of optional arguments

[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft helper functions"

Part of gh-42175. This implements the `torch.fft` helper functions: `fftfreq`, `rfftfreq`, `fftshift` and `ifftshift`.

* #43009 Cleanup tracer handling of optional arguments

[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft helper functions"

Part of gh-42175. This implements the `torch.fft` helper functions: `fftfreq`, `rfftfreq`, `fftshift` and `ifftshift`.

* #43009 Cleanup tracer handling of optional arguments

[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft helper functions"

Part of gh-42175. This implements the `torch.fft` helper functions: `fftfreq`, `rfftfreq`, `fftshift` and `ifftshift`.

* #43009 Cleanup tracer handling of optional arguments

[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft helper functions"

Part of gh-42175. This implements the `torch.fft` helper functions: `fftfreq`, `rfftfreq`, `fftshift` and `ifftshift`.

* #43009 Cleanup tracer handling of optional arguments

[ghstack-poisoned]

mruberry commented on torch/fft/__init__.py:
```diff
@@ -549,3 +555,177 @@
+fftfreq = _add_docstr(_fft.fft_fftfreq, r"""
+fftfreq(n, d=1.0, *, dtype=None, layout=torch.strided, device=None, requires_grad=False) -> Tensor
+
+Returns Discrete Fourier Transform sample frequencies for signal size :attr:`n`.
```
nit: "Computes the discrete Fourier transform sample frequencies for a signal of size :attr:\`n\`."

mruberry commented on torch/fft/__init__.py:
```diff
@@ -549,3 +555,177 @@
+
+Note:
+
+    By convention, the FFT returns positive frequency terms first, followed by
```
Should "FFT" here be a reference to to `torch.fft.fft`?

mruberry commented on torch/fft/__init__.py:
```diff
@@ -549,3 +555,177 @@
+Note:
+
+    By convention, the FFT returns positive frequency terms first, followed by
+    the negative frequencies in reverse order, so that ``f[-i]`` in python
```
What is `i` referring to? 

"python" -> "Python"

peterbell10 replied:
Clarified with "for all `i \leq n/2`".

mruberry commented on torch/fft/__init__.py:
```diff
@@ -549,3 +555,177 @@
+
+        f = [0, 1, ..., (n - 1) // 2, -(n // 2), ..., -1] / (d * n)
+
+Note:
```
This note is excellent.

mruberry commented on torch/fft/__init__.py:
```diff
@@ -549,3 +555,177 @@
+    d (float, optional): The sampling length scale.
+        The unit length (or time) between individual samples in the FFT input.
+        If given, the frequencies are given in physical units of 1/unit.
+    {dtype}
```
Keywords args:

mruberry commented on torch/fft/__init__.py:
```diff
@@ -549,3 +555,177 @@
+    n (int): the FFT length
+    d (float, optional): The sampling length scale.
+        The unit length (or time) between individual samples in the FFT input.
+        If given, the frequencies are given in physical units of 1/unit.
```
I don't think I follow what this last sentence is getting. Per the above formula, `d` just divides all the frequencies? 

Maybe there's a way to briefly explain what this does (divide the frequency samples) followed by another sentence that explains its conceptual perspective. Maybe something like: "Conceptually, the time between samples in the FFT input"?

mruberry commented on torch/fft/__init__.py:
```diff
@@ -549,3 +555,177 @@
+    >>> torch.fft.fftfreq(5)
+    tensor([ 0.0000,  0.2000,  0.4000, -0.4000, -0.2000])
+
+    In this example, the Nyquist frequency at ``f[2]`` is given as negative:
```
This suggests that the Nyquist frequency may sometimes be positive here, but per the previous note it will always be negative.

peterbell10 replied:
Changed "In this example" to "For even lengths".

mruberry commented on torch/fft/__init__.py:
```diff
@@ -549,3 +555,177 @@
+Note:
+    For even lengths, the Nyquist frequency at ``f[n/2]`` can be thought of as
+    either negative or positive. Unlike :func:`~torch.fft.fftfreq`,
+    :func:`~torch.fft.rfftfreq` always considers it to be positive.
```
Maybe: "always computes a positive value at \`\`f[n/2]\`\`."

I'm trying to work around the term "considers." (Functions are mindless, so they can't consider anything.)

mruberry commented on torch/fft/__init__.py:
```diff
@@ -549,3 +555,177 @@
+
+Note:
+
+    :func:`~torch.fft.rfft` returns Hermitian one-sided output. So, only the
```
"one-sided output. So, only " -> "one-sided output, so only

mruberry commented on torch/fft/__init__.py:
```diff
@@ -549,3 +555,177 @@
+rfftfreq = _add_docstr(_fft.fft_rfftfreq, r"""
+rfftfreq(n, d=1.0, *, dtype=None, layout=torch.strided, device=None, requires_grad=False) -> Tensor
+
+Returns sample frequencies for :func:`~torch.fft.rfft` with signal size :attr:`n`.
```
See above nit.

mruberry commented on torch/fft/__init__.py:
```diff
@@ -549,3 +555,177 @@
+
+Note:
+    By convention, the FFT returns positive frequency terms first, followed by
+    the negative frequencies in reverse order, so that ``f[-i]`` in python
```
See comment above about `i` and wording

mruberry commented on torch/fft/__init__.py:
```diff
@@ -549,3 +555,177 @@
 """)
+
+fftfreq = _add_docstr(_fft.fft_fftfreq, r"""
+fftfreq(n, d=1.0, *, dtype=None, layout=torch.strided, device=None, requires_grad=False) -> Tensor
```
n -> input to be consistent with other PyTorch ops (the other functions in this PR need the same consistency fix)

mruberry commented on torch/fft/__init__.py:
```diff
@@ -549,3 +555,177 @@
+fftshift = _add_docstr(_fft.fft_fftshift, r"""
+fftshift(x, dim=None) -> Tensor
+
+Shift zero-frequency FFT terms to the center of the tensor.
```
"Shifts the zero-frequency FFT terms to the center of :attr:\`input\`."

However, I don't know if this is a great one-sentence description of this function. It doesn't mention `dim`, and it uses the terms "shifts" and "center", which have non-obvious meanings. 

Note that "shifts" is preferable to "shift" because the function shifts things.

peterbell10 replied:
Changed to "Reorder n-dimensional FFT output to have negative frequency terms first." 

Still not 100% happy but not sure how to improve it.

mruberry commented on torch/fft/__init__.py:
```diff
@@ -549,3 +555,177 @@
+    By convention, the FFT returns positive frequency terms first, followed by
+    the negative frequencies in reverse order, so that ``f[-i]`` in python
+    gives the negative frequency terms. :func:`~torch.fft.fftshift` rearranges
+    all frequencies into ascending order from negative to positive with the
```
Wouldn't this just be sorting the tensor along the specified dim?

peterbell10 replied:
By "frequency terms" I mean the frequencies which a particular index into the tensor represents. So if `X = fft(y)` then `X[i]` is the term representing a wave of frequency `fftfreq(len(y))[i]`. 

Should I be clearer about that in the docs?

mruberry commented on torch/fft/__init__.py:
```diff
@@ -549,3 +555,177 @@
+            [-0.2000,  0.8000,  1.8000, -2.2000, -1.2000],
+            [-0.1000,  0.9000,  1.9000, -2.1000, -1.1000]])
+
+    >>> torch.fft.fftshift(x)
```
What's going on in the NumPy examples for fftshift? 

https://numpy.org/doc/stable/reference/generated/numpy.fft.fftshift.html?highlight=fftshift#numpy.fft.fftshift

mruberry replied:
I also found the NumPy example unhelpful.

mruberry commented on torch/fft/__init__.py:
```diff
@@ -549,3 +555,177 @@
+ifftshift = _add_docstr(_fft.fft_ifftshift, r"""
+ifftshift(x, dim=None) -> Tensor
+
+Inverse of :func:`fftshift`.
```
Using "inverse" is OK here because this sentence is a description of the function.

mruberry commented on torch/fft/__init__.py:
```diff
@@ -549,3 +555,177 @@
+
+Inverse of :func:`fftshift`.
+
+This takes the FFT representation with zero-frequency FFT terms in the center
```
I'll take another look at this documentation after the other updates (and consistency updates).

mruberry replied:
Is "FFT representation" defined somewhere? 

I actually wonder on this function how much more description than "inverse of fftshift" is needed. Is there a use case for this other than inverting the fftshift results?

mruberry commented on torch/csrc/api/include/torch/fft.h:
```diff
@@ -166,4 +166,65 @@ inline Tensor ihfft(const Tensor& self,
   return torch::fft_ihfft(self, n, dim, norm);
 }
 
+/// Returns Discrete Fourier Transform sample frequencies for signal size n.
```
These comments are great but may need a small consistency update if the one-sentence descriptions in _torch_docs.py change.

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -7807,6 +7807,26 @@
+  use_c10_dispatcher: full
+  variants: function
+
+- func: fft_fftshift(Tensor x, int[1]? dim=None) -> Tensor
```
`Tensor self` here and "input" in the docs for consistency.

mruberry commented on torch/fft/__init__.py:
```diff
@@ -549,3 +555,177 @@
+""".format(**factory_common_args))
+
+fftshift = _add_docstr(_fft.fft_fftshift, r"""
+fftshift(x, dim=None) -> Tensor
```
"x" -> "input" for consistency with other PyTorch ops

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -7807,6 +7807,26 @@
   use_c10_dispatcher: full
   variants: function
 
+- func: fft_fftfreq(int n, float d=1.0, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
```
would "window_length" be more consistent with the existing fft functions, as opposed to "n" as the first argument?

peterbell10 replied:
`n` is the name of the second argument to `fft`. So, I think that in this case being consistent with numpy is more important.

mruberry commented on aten/src/ATen/native/SpectralOps.cpp:
```diff
@@ -409,6 +409,66 @@ Tensor fft_irfftn(const Tensor& self, c10::optional<IntArrayRef> s,
+  Tensor result = native::arange(n, options);
+  auto right_slice = result.slice(0, (n + 1) / 2, 0);
+  at::arange_out(right_slice, -(n/2), 0, 1);
+  result.mul_(1.0 / (n * d));
```
Maybe a brief note here explaining why the multiplication by the reciprocal is occurring instead of a call to `true_divide_`.

mruberry commented on aten/src/ATen/native/SpectralOps.cpp:
```diff
@@ -409,6 +409,66 @@ Tensor fft_irfftn(const Tensor& self, c10::optional<IntArrayRef> s,
+  return result;
+}
+
+DimVector default_alldims(const Tensor& self, c10::optional<IntArrayRef> dim_opt) {
```
Brief comment describing this function

mruberry commented on test/test_spectral_ops.py:
```diff
@@ -506,6 +506,78 @@ def test_fftn_invalid(self, device):
+    # Helper functions
+
+    @skipCPUIfNoMkl
+    @skipCUDAIfRocm
```
XLA build is currently failing for unrelated reasons but you may need to skip this test on XLA (by testing it only on CPU and CUDA backends).

mruberry replied:
(same for the following tests)

mruberry commented on test/test_spectral_ops.py:
```diff
@@ -506,6 +506,78 @@ def test_fftn_invalid(self, device):
+    @skipCPUIfNoMkl
+    @skipCUDAIfRocm
+    @unittest.skipIf(not TEST_NUMPY, 'NumPy not found')
+    @dtypes(torch.float, torch.double)
```
Why no complex dtypes here?

peterbell10 replied:
`fftfreq` relies on `arange` which isn't implemented for complex.

mruberry replied:
Maybe add a TODO to the cpp implementation?

peterbell10 committed:
Update on "torch.fft helper functions"

Part of gh-42175. This implements the `torch.fft` helper functions: `fftfreq`, `rfftfreq`, `fftshift` and `ifftshift`.

* #43009 Cleanup tracer handling of optional arguments

[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft helper functions"

Part of gh-42175. This implements the `torch.fft` helper functions: `fftfreq`, `rfftfreq`, `fftshift` and `ifftshift`.

* #43009 Cleanup tracer handling of optional arguments

[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft helper functions"

Part of gh-42175. This implements the `torch.fft` helper functions: `fftfreq`, `rfftfreq`, `fftshift` and `ifftshift`.

* #43009 Cleanup tracer handling of optional arguments

[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft helper functions"

Part of gh-42175. This implements the `torch.fft` helper functions: `fftfreq`, `rfftfreq`, `fftshift` and `ifftshift`.

* #43009 Cleanup tracer handling of optional arguments

[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft helper functions"

Part of gh-42175. This implements the `torch.fft` helper functions: `fftfreq`, `rfftfreq`, `fftshift` and `ifftshift`.

* #43009 Cleanup tracer handling of optional arguments

[ghstack-poisoned]

peterbell10 commented:
@mruberry PTAL.

mruberry commented on torch/fft/__init__.py:
```diff
@@ -549,3 +555,186 @@
+Note:
+    For even lengths, the Nyquist frequency at ``f[n/2]`` can be thought of as
+    either negative or positive. Unlike :func:`~torch.fft.fftfreq`,
+    :func:`~torch.fft.rfftfreq` always returns is as positive.
```
"returns is as positive" -> "returns it as positive"

mruberry commented on torch/fft/__init__.py:
```diff
@@ -549,3 +555,186 @@
+fftshift = _add_docstr(_fft.fft_fftshift, r"""
+fftshift(input, dim=None) -> Tensor
+
+Reorder n-dimensional FFT output to have negative frequency terms first. 
```
"Reorder" -> "Reorders"

mruberry replied:
Would it be helpful to say "first by ..." with a description of the rolling the function performs? The current one-sentence descriptor suggests the function has some intelligence. Maybe that's fine.

mruberry commented on torch/fft/__init__.py:
```diff
@@ -549,3 +555,186 @@
+Note:
+    For even lengths, the Nyquist frequency at ``f[n/2]`` can be thought of as
+    either negative or positive. Like :func:`~torch.fft.fftfreq`,
+    :func:`~torch.fft.fftshift` always considers it to be negative.
```
This note is a little confusing because fftshift doesn't actually look at its input. Isn't it more like the function requires this to work properly?

mruberry replied:
Maybe this should just be a note about how this function really just performs a particular kind of roll, so its input needs to be structured like fftshift's output?

peterbell10 replied:
>Isn't it more like the function requires this to work properly?

No, the Nyquist term will always exist in the input. This note is describes where the Nyquist term gets placed in the return value.

Because of the sampling theorem, the Nyquist frequency term in both positive and negative frequencies are identical. So, what changes is whether you include the Nyquist term at the start (treating it as negative), or at the end (treating it as positive). Mathematically, these are equally valid choices.

mruberry replied:
I understand that. But isn't the point that if the term was positive then the operation performed by fftshift wouldn't reorder it properly?

peterbell10 replied:
>But isn't the point that if the term was positive then the operation performed by fftshift wouldn't reorder it properly?

I'm not quite sure what you mean. The point it that `fftshift` puts the Nyquist term in the left-most position, which corresponds to a negative frequency.

mruberry replied:
Right. Actually being explicit that fftshift just always puts that term at the zero index position would be good. The point I'm trying to get at is the function just always does that, regardless of the actual values. And so for the function fulfill its conceptual purpose, that value should be negative.

Does that make sense?

We could also have a brief VC to discuss this. Sometimes text isn't the greatest way to communicate.

peterbell10 replied:
>Right. Actually being explicit that fftshift just always puts that term at the zero index position would be good. 

I updated the wording to "fftshift always puts the Nyquist term at the 0-index; equivalent to the negative frequency. This is the same convention used by fftfreq."

>The point I'm trying to get at is the function just always does that, regardless of the actual values. And so for the function fulfill its conceptual purpose, that value should be negative.

I don't understand this point. The FFT gives amplitudes of waves, not their frequencies. The value of the Nyquist term can be anything and does not depend on your convention for the sign of the frequency.

mruberry replied:
I get your point. Maybe we should briefly vc about it? Slack me sometime that's convenient for you?

mruberry commented on torch/fft/__init__.py:
```diff
@@ -549,3 +555,186 @@
+    input (Tensor): the tensor in FFT order
+    dim (int, Tuple[int], optional): The dimensions to rearrange.
+        Only dimensions specified here will be rearranged, any other dimensions
+        will be left in their original order.
```
"The dimensions to rearrange. If None, ... . If not None, only the specified dimensions will be rearranged. Default is None."

peterbell10 replied:
Not mentioning `None` explicitly is consistent with the other docs I've written for this module. Is that not okay?

mruberry replied:
It's OK. We can do an overall consistency pass later.

mruberry commented on torch/fft/__init__.py:
```diff
@@ -549,3 +555,186 @@
+    >>> torch.fftshift(f)
+    tensor([-0.5000, -0.2500, 0.0000, 0.2500])
+
+    Also notice that the Nyquist frequency at `f[2]` was considered negative
```
This aside will likely require a consistency update.

mruberry commented on torch/fft/__init__.py:
```diff
@@ -549,3 +555,186 @@
+
+Args:
+    input (Tensor): the tensor in FFT order
+    dim (int, Tuple[int], optional): The dimensions to rearrange.
```
See previous comment.

mruberry commented on torch/fft/__init__.py:
```diff
@@ -549,3 +555,186 @@
+    >>> f
+    tensor([ 0.0000,  0.2000,  0.4000, -0.4000, -0.2000])
+
+    A round-trip through :func:`~torch.fft.fftshift` and
```
Nice.

peterbell10 committed:
Update on "torch.fft helper functions"

Part of gh-42175. This implements the `torch.fft` helper functions: `fftfreq`, `rfftfreq`, `fftshift` and `ifftshift`.

* #43009 Cleanup tracer handling of optional arguments

[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft helper functions"

Part of gh-42175. This implements the `torch.fft` helper functions: `fftfreq`, `rfftfreq`, `fftshift` and `ifftshift`.

* #43009 Cleanup tracer handling of optional arguments

[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft helper functions"

Part of gh-42175. This implements the `torch.fft` helper functions: `fftfreq`, `rfftfreq`, `fftshift` and `ifftshift`.

* #43009 Cleanup tracer handling of optional arguments

[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft helper functions"

Part of gh-42175. This implements the `torch.fft` helper functions: `fftfreq`, `rfftfreq`, `fftshift` and `ifftshift`.

* #43009 Cleanup tracer handling of optional arguments

[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft helper functions"

Part of gh-42175. This implements the `torch.fft` helper functions: `fftfreq`, `rfftfreq`, `fftshift` and `ifftshift`.

* #43009 Cleanup tracer handling of optional arguments

[ghstack-poisoned]

peterbell10 commented:
@mruberry updated the `fftshift` wording. Also cleaned out a few unnecessary blank lines and rebased on `viable/strict` to give CI a chance to run.

mruberry commented:
Hey @peterbell10 Thanks for the updates. I'll take a look later this week after the 1.7 branch cut.

mruberry reviewed (approved):
Cool!

Branch is cut so back to work on reviews. Also I expect to update the fft roll-up issue soon. test_fft.py is taking a significant amount of time at the moment, so it'll be good to put performance analysis on the calendar. Looking forward to getting your thoughts on possible future directions, too, @peterbell10!

peterbell10 committed:
Update on "torch.fft helper functions"

Part of gh-42175. This implements the `torch.fft` helper functions: `fftfreq`, `rfftfreq`, `fftshift` and `ifftshift`.


Differential Revision: [D24043473](https://our.internmc.facebook.com/intern/diff/D24043473)

[ghstack-poisoned]

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1398,6 +1398,12 @@ def method_tests():
         ('__getitem__', torch.randn(S, S, S), (dont_convert([[0, 2, 3], [1, 3, 3],
                                                              torch.LongTensor([0, 0, 2])]),), 'adv_index_var'),
         ('to_sparse', (S, S), (), '', (), (), [], lambda x: x.to_dense()),
+        ('fft.fftshift', (S,), NO_ARGS, '1d', (), (None, 0, -1)),
```
Looks like we'll need to remove these as the internal tests are complaining yet again. I'm not too worried, however. The OSS CI has run them and the autograd formulas are unlikely to change. In the near future we'll convert them into OpInfos and test their autograds that way, too.

peterbell10 committed:
Update on "torch.fft helper functions"

Part of gh-42175. This implements the `torch.fft` helper functions: `fftfreq`, `rfftfreq`, `fftshift` and `ifftshift`.


Differential Revision: [D24043473](https://our.internmc.facebook.com/intern/diff/D24043473)

[ghstack-poisoned]

mruberry commented:
@peterbell10 Looks like these updates aren't appearing in the documentation because we forgot to update the fft.rst file. Would you submit a PR adding them?
